### PR TITLE
Update firestore externs file to avoid closure compiler errors

### PR DESCRIPTION
**Pull request by hsubox76**
Upon importing `firebase-firestore-externs.js` for internal use for the first time, some errors came up when a project attempted to build it using closure compiler, mostly regarding unrecognized jsdoc annotations. Fixed errors brought up by the compiler.

### CODE REVIEW

File: packages/firebase/externs/firebase-firestore-externs.js [hunk 1]
```diff
@@ -712,42 +714,45 @@ firebase.firestore.SnapshotMetadata.prototype.isEqual = function(other) {};
  * access will return 'undefined'. You can use the `exists` property to
  * explicitly verify a document's existence.
  *
- * @interface
+ * @constructor
```
Comment (mikelehen) [inline packages/firebase/externs/firebase-firestore-externs.js]: Why did this change?  We don't allow creating DocumentSnapshot objects directly...  and I see we (successfully?) use `@constructor` elsewhere in this file...

Comment (mikelehen) [reply packages/firebase/externs/firebase-firestore-externs.js]: woops, I meant we successfully use `@interface` elsewhere in this file.

Comment (hsubox76) [reply packages/firebase/externs/firebase-firestore-externs.js]: The closure compiler error was "constructors can only extend constructors" when `QueryDocumentSnapshot` tried to extend this.

**Comment by hsubox76** [comment]
Figuring out a way to generate this from index.d.ts might be a good idea for the future, especially if we are going to be expected to keep the third_party version updated regularly. The new Typedoc autodocumentation tools might help somehow.

**Comment by Feiyang1** [comment]
They are used by 1p apps as dependencies for their compilation.
We have another set of externs in another location in google3 that are used for reference doc generation. Unfortunately it is also used as dependency in some 1p projects. So we can't remove either of them without breaking something.

Two things I think we can do:
1. 1p projects should really only take dependency on firebase externs in third_party/firebase, so we can try to change the 1p projects to depend on the right externs and we can remove the externs in firebase/jscore/api.

2. There is a tool called Tsickle that compiles d.ts file to externs. I have tried it before in the attempt to create externs files and use our existing jsdoc scripts to create reference docs. Not sure if the output externs can be directly used for other projects, but It's worth exploring.

**Comment by mikelehen** [comment]
Cool, thanks!

FWIW- If there are only a few 1p projects using these and they're relatively slow-moving (don't rely on our newest features), it's tempting to push the burden of maintaining externs onto them...  just as they would have to do for any other non-closure dependencies that they're using.

Basically I'd assert that closure usage should be going down over time, and we should focus on supporting TypeScript.  We can't break people, but we might be able to move towards deprecating first-class closure support.

Of course if we can support them for free via tsickle generation or something, then this becomes a non-issue.

File: packages/firebase/externs/firebase-firestore-externs.js (hunk 1 above)
Comment (mikelehen) [reply packages/firebase/externs/firebase-firestore-externs.js]: I see.  Perhaps the better fix is to change QueryDocumentSnapshot to use`@interface` then.  I don't care super strongly about these externs being perfectly correct though. :)

**Pull request merged**